utils/autosp/autosp-src/autosp.1
.\" This manpage is licensed under the GNU Public License
.TH AUTOSP 1 2023-11-26  "" ""

.SH NAME
autosp \- preprocessor to generate note-spacing commands for MusiXTeX scores

.SH SYNOPSIS
.B autosp 
.RB  [ \-v\ |\ \-\-version\ |\ \-h\ |\ \-\-help ]

.B autosp
.RB [ \-d\ |\ \-\-dotted ]
.RB [ \-l\ |\ \-\-log ]
.IR infile [\fR.aspc\fP]
.RI [ outfile [ \fR.tex\fP ]]

.SH DESCRIPTION
This program makes it easier to create MusiXTeX scores by converting (non-standard) commands of the 
form \\anotes ... \\en (or \\an ... \\en) into one or more conventional
note-spacing commands (\\notes \\Notes \\NOtes ...), 
determined by the actual note values, with \\sk spacing commands inserted as necessary.
The coding
for an entire measure can be entered one part at a time, 
without
concern for note-spacing changes within the part or spacing requirements of other parts.

For example, if applied to
.nf

    \\anotes\\qa J\\qa K&\\ca l\\qa m\\ca n\\en 

.fi
.B autosp
generates
.nf

    \\Notes\\qa J\\sk\\qa K\\sk&\\ca l\\qa m\\sk\\ca n\\en

.fi
Typically,
an
\\anotes command generates several conventional
note-spacing commands.

If the 
.I infile
argument 
does not have an .aspc extension, input is taken from 
.IR infile .aspc 
if that file exists.

If the
.I outfile
argument does not have a .tex extension, output is sent to
.IR outfile .tex.
If no 
.I outfile 
argument is provided, output will go to
.IR infile .tex
(or to
.IR infile .ltx
if a \\documentclass declaration is encountered).

For
\\anotes commands,
line breaks and spaces may precede note segments, allowing more flexible source formatting; 
the line breaks and spaces will be elided from the output. 

For example, 
.nf

  \\an
    \\ibl0K0\\qb0K\\nbbl0\\qb0{.K}\\tbbbl0\\tbbl0\\tql0L&
    \\ibbl1m{-2}\\qb1{.m}\\tbbbl1\\tbbl1\\qb1l\\tql1k\\en

.fi
is acceptable and generates
.nf

  \\notesp\\ibl0K0\\qb0K&\\ibbl1m{-2}\\qb1{.m}\\en
  \\nnotes\\sk&\\tbbbl1\\tbbl1\\qb1l\\en
  \\notesp\\nbbl0\\qb0{.K}&\\tql1k\\en
  \\nnotes\\tbbbl0\\tbbl0\\tql0L&\\sk\\en

.fi


If the 
.B \-l 
( \fB\-\-log\fP ) option is used, a very detailed log 
.IR infile .alog
is generated.

If the 
.B \-d  
( \fB--dotted\fP )
option is used, 
.I dotted 
beam notes of the form \\qb{\fIn\fP}{\fB.\fP\fIp\fP} are 
.I not 
given extra space;
it is assumed that the subsequent note will be shifted
by 
a \\roff-like command 
or a spacing command such as \\qsk or \\hqsk.
Commands of the form \\qlp{\fIp\fP}, \\qlpp{\fIp\fP}, ..., \\qpb{\fIn\fP}{\fIp\fP} and 
\\qppb{\fIn\fP}{\fIp\fP} are always spaced
as indicated. 

If there is a single staff, consecutive whole-bar rest bars
are merged into a multi-bar rest. Bar-centered rests can be
coded using the standard \\def\\atnextbar notation but 
the non-standard command \\Cpause (or \\Cp) in a note segment also generates a
bar-centered rest.

Spacing commands \\sk and \\bsk in the source are discarded,
but not "small" skips \\hsk, \\qsk, \\tqsk, \\hqsk or \\qqsk,
or the small "backward" skips \\hbsk, \\bqsk, \\btsk, or \\bhsk.
Moreover,
non-standard commands
\\QQsk, \\HQsk \\TQsk and \\Qsk in the source generate "global" skips; i.e., the effect of \\qqsk, \\hqsk, \\tqsk or \\qsk, respectively,  in 
.I every 
staff. These ensure that staffs remain synchronized if additional spacing is needed in any staff(s). 

Global skips may also be obtained within collective-coding sequences by using
up to four successive commas 
to get the effects of \\QQsk, \\HQsk, \\TQsk or \\Qsk, respectively.
Global skips \\tqsk (or, for double-flats, \\qsk) are 
automatically inserted before accidentals ('^', '_', '=', '<', '>') on collective-coding notes
(except when the preceding note is "virtual"; i.e., a skip).
If this automatic additional spacing is 
.I not 
wanted in some context, it may be avoided by
replacing 
the accidental in the collective-coding sequence
by 
any of the 
.I explicit 
accidental commands: \\sh, \\fl, \\na, \\smallsh, \\bigsh, etc.
If the automatic spacing is insufficient, the spacing may be increased by adding sufficient commas or 
using a conventional notes command instead of \\an.  


A note segment can be completely empty, but if a note segment should start
with or contain a "space," the note-value
of that space must be made explicit with a command of the
form \\ha{*}, \\qa{.*}, \\qa{*}, \\ca{*}, etc. 

From version 2017-06-14, the effects of \\TransformNotes calls are implemented by the 
.B autosp
pre-processor. This
enables use of 
musixlyr in
.B autosp 
scores; musixlyr.tex 
is incompatible with the
musixtnt.tex 
implementation of \\TransformNotes.

All other conventional MusiXTeX commands are output exactly as given in the input.


.SH OPERATION

.B autosp
determines the spacing for ordinary notes from the note
commands themselves; for example, 
.TP 3
+
\\qa, \\qu, \\ql, \\qp result in
\\NOtes; 
.TP 3
+
\\ca, \\cu, \\cl, \\ds result in \\Notes; 
.PP
and so on. 
 
The spacing for 
.IR beamed 
notes is determined by the beam
multiplicity, so that \\ib... results in \\Notes, \\ibb... results in
\\notes, etc. 

Collective coding of note sequences (including accidentals and dots) is handled 
by expanding the sequence
into a sequence of individual note commands.


.B autosp 
scales notes by the relevant instrument-size value.
 
.SH LIMITATIONS

.B autosp
assumes that & and | (rather than \\nextinstrument and \\nextstaff) are used to separate instruments and staffs.


Appoggiaturas and grace notes are recognized by the use of \\tinynotesize; note-spacing
of 1.45\\elemskip is used. If this isn't suitable and can't be corrected with a small skip, a
\\vnotes command with any desired spacing can be used.

.B autosp 
supports 
.I x-tuplets 
introduced using \\xtuplet{\fIx\fP}
or \\xxtuplet{\fIx:y\fP}
and 
.I triplets 
introduced
using any of the following commands (regardless of any re-definition of \\txt or \\tuplettxt):
.nf

  \\triolet
  \\uptrio
  \\downtrio
  \\uptuplet
  \\downtuplet

.fi
.B autosp 
assumes that an x-tuplet is to be played in 
(\fIx\fP\-1)/\fIx\fP 
of the apparent x-tuplet duration.
So, for example, a triplet in eighths is assumed to be played in the time of one quarter note.
If this assumption
isn't valid, the x-tuplet must
be coded explicitly using a suitable \\vnotes command; see the first
measure of barsant2.aspc for an example of a non-standard x-tuplet: a 5-tuple 
of 64th notes with an intended duration
of 
.I six 
64ths.

In some polyrhythmic scores, the \\txt numeral may be displaced, even if the notes
themselves are correctly spaced. In these cases, it is possible to suppress 
the normal output of \\txt by using the non-standard commands \\Triolet (no arguments)
or \\Xtuplet{k}\\
and placing a numeral at the correct location using \\zcn (i.e., \\zcharnote).

.B autosp
can deal with simultaneous x-tuplets in multiple staffs provided the x values and
total note durations are identical.  

In some baroque scores, particularly by J.S. Bach, a beamed sixteenth note is vertically aligned
with
the third note of a triplet of eighth notes in another staff (implying that they should
be sounded simultaneously); e.g.,
.nf

    \\ibl0L0\\qb0{.L}\\tqql0L

.fi
would be played as if notated 
.nf

    \\uptrio{b}10\\ql L\\hroff{\\cl L}

.fi
The following coding will align the beamed sixteenth note 
with the third note of a triplet in another staff:
.nf

    \\ibl0L0\\qb0{.L}\\hbsk\\tqql0L

.fi
and, similarly, for triplets of sixteenth notes:
.nf

    \\ibbu0J0\\qb0{.J}\\hbsk\\nqqqu0J\\qb0{.J}\\hbsk\\tqqqu0J

.fi
Generally, user-defined macros are not processed or expanded; however, definitions of the form
.nf

    \\def\\atnextbar{\\znotes ... \\en}

.fi
generate definitions that do take account of \\TransformNotes.

All staffs are assumed to have the same meter; see kinder2.aspc for
an example of how to work around this.

.B autosp
may not be effective for music with more than one voice in a single staff. It might be
advisable to use a separate staff for each voice, to avoid \\anotes when necessary, 
or to omit certain voices
initially and add them into the resulting TeX file.

.SH EXAMPLES
See files quod2.aspc, kinder2.aspc, geminiani.aspc and barsant2.aspc for scores suitable for input to  
.BR autosp .
The program
.B tex2aspc
can be used to convert "legacy" MusiXTeX scores to .aspc format.

.SH SEE ALSO
.BR msxlint (1)
.BR tex2aspc (1)
.PP 
musixdoc.pdf

.SH AUTHOR 
This program and manual page were written by Bob Tennent <rdt@cs.queensu.ca>.


ChangeLog
2025-10-16  Karl Berry  <karl@freefriends.org>

	* Build (status): avoid using egrep.
	Report from Jan Kasprzak,
	https://tug.org/pipermail/tex-k/2025-October/004217.html.

2025-09-07  Karl Berry  <karl@freefriends.org>

	* Makefile.am (install_target): new variable, install-strip by default.
	(world): use it.

2025-07-22  Karl Berry  <karl@freefriends.org>

	* reautoconf: mention that autoconf-only packages need to be
	explicitly added; would be better to read kpse-pkgs.m4.
	Also reautoconf tree for kpse-setup.m4 changes.

2025-06-26  Karl Berry  <karl@freefriends.org>

	* automake 1.18.1.

2025-04-30  Karl Berry  <karl@freefriends.org>

	* Build: incorporate original CFLAGS, CXXFLAGS, OBJCXXFLAGS values.
	https://tug.org/pipermail/tlbuild/2025q2/005771.html

2025-04-29  Karl Berry  <karl@freefriends.org>

	* Build: use TL_CFLAGS, TL_CXXFLAGS, etc., for TL_BUILD_ENV
	if -g is not given.

2025-03-08  Karl Berry  <karl@tug.org>

	* version.ac: 2026/dev.

2025-03-07  Karl Berry  <karl@tug.org>

	* TL'25 release.

2025-01-12  Karl Berry  <karl@tug.org>

	* Build: also support --debug-more for -g -Og -ggdb3,
	per Luigi, as described on https://tug.org/texlive/build.html.

2024-07-12  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake 1.17.

2024-04-29  Karl Berry  <karl@freefriends.org>

	* Makefile.am (distdir): override default dist generation,
	since we don't try to keep make dist working any more.
	https://tug.org/pipermail/tlbuild/2024q2/005547.html

2023-12-31  Karl Berry  <karl@freefriends.org>

	* reautoconf: add -f / --force option to pass to autoreconf.

2023-12-26  Karl Berry  <karl@freefriends.org>

	* reautoconf: don't bother with environment variables;
	because the autotools have many subprograms, setting PATH
	is what makes sense.

2023-12-23  Karl Berry  <karl@tug.org>

	* reautoconf with autoconf-2.72.

2023-12-21  Karl Berry  <karl@freefriends.org>

	* reautoconf (do_it): abort the run if any command fails.

2023-02-25  Karl Berry  <karl@tug.org>

	* Build: show full uname -a.

2023-02-09  Karl Berry  <karl@freefriends.org>

	* version.ac: 2023, no more /dev.

2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2022-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): let's hope for 2019-04-07.

2018-10-02  Karl Berry  <karl@tug.org>

	* Build: TL_MAKE_VERBOSE=VERBOSE=1 by default,
	so that test-suite.log will be shown on failure.
	https://lists.gnu.org/archive/html/automake/2018-09/msg00015.html

2018-04-28  Karl Berry  <karl@tug.org>

	* version.ac: 2019/dev.

2018-04-10  Karl Berry  <karl@tug.org>

	* README.*: update from new tlbuild.

2017-11-15  Karl Berry  <karl@tug.org>

	* *.test: use #!/bin/sh -vx. Add 

2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the

utils/autosp/autosp-src/autosp.c
char version[12] = "2020-02-06";

/*  Copyright (C) 2014-20 R. D. Tennent School of Computing,
 *  Queen's University, rdt@cs.queensu.ca
 *
 *  This program is free software; you can redistribute it
 *  and/or modify it under the terms of the GNU General
 *  Public License as published by the Free Software
 *  Foundation; either version 2 of the License, or (at your
 *  option) any later version.
 *
 *  This program is distributed in the hope that it will
 *  be useful, but WITHOUT ANY WARRANTY; without even the
 *  implied warranty of MERCHANTABILITY or FITNESS FOR A
 *  PARTICULAR PURPOSE. See the GNU General Public License
 *  for more details.
 *
 *  You should have received a copy of the GNU General
 *  Public License along with this program; if not, write to
 *  the Free Software Foundation, Inc., 51 Franklin Street,
 *  Fifth Floor, Boston, MA 02110-1301, USA.
 */

/*  autosp - preprocessor to generate note-spacing commands for MusiXTeX scores */

/*  Usage: autosp [-v | --version | -h | --help]
 *         autosp [-d | --dotted] [-l | --log] infile[.aspc] [outfile[.tex]]
 *
 *  Options 
 *    --dotted (-d) suppresses extra spacing for dotted beam notes.
 *    --log (-l) generate a log file
 *
 *
 *
 *  Key functions:
 *
 *    process_command, which processes commands
 *      \instrumentnumber
 *      \nbinstrument
 *      \setstaffs
 *      \startpiece
 *      \TransformNotes
 *      \def\vnotes
 *      \def\atnextbar
 *      \anotes etc.
 *      \bar
 *      \endpiece etc.
 *
 *    analyze_notes, which determines pointers notes[i] and 
 *      characters terminator[i] for each staff; '$' is used
 *      as the terminator for the last staff.
 *
 *    generate_notes, which repeatedly uses spacing_note (i) to
 *      determine spacings[i] of the next spacing note in 
 *      each staff; the current spacing is then the smallest
 *      of these.  If necessary, the current notes command
 *      is terminated by terminate_notes() and a new notes 
 *      command initiated by initialize_notes(). The current global 
 *      skip and note commands for each staff are then copied to 
 *      outstrings[i] using output_notes (i).  When necessary, 
 *      vspacing[i] is set to generate subsequent skips.
 *      Appoggiaturas and xtuplets are treated separately.
 *
 *    spacing_note (i), which searches for the next spacing note
 *      in the i'th staff, unless a "virtual note" (skip) is
 *      needed.
 *  
 *    output_notes (i), which copies a global_skip string and 
 *    outstrings[i] to output, using filter_output (i) to discard
 *    commas and global and large skips, and insert collective-coding 
 *    prefixes and closing braces.  
 *
 *    terminate_notes, which copies outstrings[i] to output,
 *    in the order determined by TransformNotes2.
 *
 *  Spacing values are expressed as sums of binary multiples of 
 *    SMALL_NOTE (by default, 1/256th); for example, half-note
 *    spacing = SP(2) = SMALL_NOTE/2. Special small or large 
 *    values are used for appoggiaturas and bar-rests.
 *      
 */

# include "utils.h"

#ifdef HAVE_CONFIG_H   /* for TeXLive */
#include <config.h>
#endif

# ifdef KPATHSEA
# include <kpathsea/getopt.h>
# else
# include <getopt.h>
# endif

# ifdef HAVE_STDBOOL_H
# include <stdbool.h>
# else
# ifndef HAVE__BOOL
# define _Bool signed char
# endif
# define bool    _Bool
# define true    1
# define false   0
# endif

# define PRIVATE static

# define SHORT_LEN 256

bool debug;
bool dottedbeamnotes;          /* dotted beam notes ignored?        */

char infilename[SHORT_LEN];
char *infilename_n;
char outfilename[SHORT_LEN];
char *outfilename_n;
char logfilename[SHORT_LEN];
char *logfilename_n;
FILE *infile, *outfile, *logfile;


void process_score ();

PRIVATE void
usage ()
{
  printf ("Usage: autosp [-v | --version | -h | --help]\n");
  printf ("       autosp [-d | --dotted] [-l | --log] infile[.aspc] [outfile[.tex]]\n");
  printf ("       autosp [-d | --dotted] [-l | --log] infile.tex outfile[.tex]\n");
}


bool dottedbeamnotes = false;   /* dotted beam notes ignored?        */
bool debug = false;

int main (int argc, char *argv[])
{
  int c;
# define NOPTS 5
  struct option longopts[NOPTS] =
  {  { "help", 0, NULL, 'h'},
     { "version", 0, NULL, 'v'},
     { "dotted", 0, NULL, 'd'},
     { "log", 0, NULL, 'l'},
     { NULL, 0, NULL, 0}
  };
  
  printf ("This is autosp, version %s.\n", version);
  printf ("Copyright (C) 2014-20  R. D. Tennent\n" );
  printf ("School of Computing, Queen's University, rdt@cs.queensu.ca\n" );
  printf ("License GNU GPL version 2 or later <http://gnu.org/licences/gpl.html>.\n" );
  printf ("There is NO WARRANTY, to the extent permitted by law.\n\n" );

  c = getopt_long (argc, argv, "hvdl", longopts, NULL);
  while (c != -1)
    {
      switch (c)
        {
        case 'h':
          usage ();
          printf ("Please report bugs to rdt@cs.queensu.ca.\n" );
          exit (0);
        case 'v':
          exit (0);
        case 'd':
          dottedbeamnotes = true;
          break;
        case 'l':
          debug = true;
          break;
        case '?':
          exit (EXIT_FAILURE);
        default:
          printf ("Function getopt returned character code 0%o.\n",
                  (unsigned int) c);
          exit (EXIT_FAILURE);
        }
      c = getopt_long (argc, argv, "hvd", longopts, NULL);
    }

  infilename[0] = '\0';
  infilename_n = infilename;
  if (optind < argc)
  { append (infilename, &infilename_n, argv[optind], sizeof (infilename));
    if (!suffix (".aspc", infilename)) 
      append (infilename, &infilename_n, ".aspc", sizeof (infilename));
  }
  else 
  {  usage();
     exit (EXIT_FAILURE);
  }
  infile = fopen (infilename, "r");
  if (infile == NULL )
  { printf ("Can't open %s\n", infilename);
    exit (EXIT_FAILURE);
  }

  printf ("Reading from %s.", infilename);

  optind++;
  outfilename[0] = '\0';
  outfilename_n = outfilename;
  if (optind < argc)  /* user-provided outfilename */
  {
    append (outfilename, &outfilename_n, argv[optind], sizeof (outfilename));
    if (!suffix (".tex", outfilename))
      append (outfilename, &outfilename_n, ".tex", sizeof (outfilename));
  }
  else if (suffix (".aspc", infilename)  )
  { /* output to infile.tex  */
    append (outfilename, &outfilename_n, infilename, sizeof (outfilename));
    outfilename_n -= 4;
    *outfilename_n = '\0';
    append (outfilename, &outfilename_n, "tex", sizeof (outfilename));
  }

  if (*outfilename == '\0')
  { 
    printf ("\nPlease provide a filename for the output.\n\n");
    exit (EXIT_FAILURE);
  }
  else
  {
    outfile = fopen (outfilename, "wb");
    if (outfile == NULL)
    { printf ("Can't open %s\n", outfilename);
      exit (EXIT_FAILURE);
    }
    printf (" Writing to %s.", outfilename);
  }

  logfilename[0] = '\0';
  logfilename_n = logfilename;
  append (logfilename, &logfilename_n, infilename, sizeof (logfilename));
  if (suffix (".aspc", infilename) )
    logfilename_n -= 4;
  else
    logfilename_n -= 3;
  *logfilename_n = '\0';
  append (logfilename, &logfilename_n, "alog", sizeof (logfilename));
  if (debug)  
  { logfile = fopen (logfilename, "wb");
    if (logfile == NULL)
    { printf ("Can't open %s\n", logfilename);
      exit (EXIT_FAILURE);
    }
    printf (" Log file %s.", logfilename);
  }
  printf ("\n\n");

  fprintf (outfile, "%%  Generated by autosp (%s).\n", version);

  process_score ();
  
  return 0;
}
